test/testdata/desugar/sclass.rb

```diff
@@ -61,10 +61,12 @@ class G
     def wrapper
         class << self
             def inner
-                T.reveal_type(self) # error: type: `T.class_of(G)`
+                T.reveal_type(self) # error: type: `G`
             end
         end
-        inner # error: Method `inner` does not exist on `G`
+        inner # this would not be allowed dynamically, but we treat
+              # methods defined within instance methods as instance
+              # methods
```

Comment:
so it seems like we ignore the `class << self` inside the def? why?

Reply:
A `class << self` inside an instance method opens the singleton class of the instance, which is not a thing we have a way of modeling. However, the resulting method _is_ valid on instances, but not on _every_ instance, so this is just an overapproximation.
```ruby
irb(main):001:0> class A
irb(main):002:1>   def foo
irb(main):003:2>     class << self
irb(main):004:3>       def bar; end
irb(main):005:3>     end
irb(main):006:2>   end
irb(main):007:1> end
=> :foo
irb(main):008:0> instance = A.new
=> #<A:0x00007ff0a00a6188>
irb(main):009:0> instance.foo
=> :bar
irb(main):010:0> instance.bar  # <----
=> nil
irb(main):011:0> A.new.foo
=> :bar
irb(main):012:0> A.new.bar  # <----
Traceback (most recent call last):
        4: from /Users/gdritter/.rbenv/versions/2.6/bin/irb:23:in `<main>'
        3: from /Users/gdritter/.rbenv/versions/2.6/bin/irb:23:in `load'
        2: from /Users/gdritter/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        1: from (irb):12
NoMethodError (undefined method `bar' for #<A:0x00007ff0a00b5048>)
```